Bump DefaultKubeBinaryVersion to 1.33

Jefftree opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

Bump DefaultKubeBinaryVersion to 1.33

- Fixed the etcd tests to account for previous emulated versions
- Removed a emulated version test that tests for 1.30. Unfortunately we don't have a GA API in the 1.31-1.32 range to test the storage version, it might be worthwhile to (similar to the metrics test) create a fake API that we inject into tests for this purpose.
- Changed the version validation to support n-3 with clamp at 1.31 min for emulatedVersion and 1.30 min for minCompatibilityVersion

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

/triage accepted

Jefftree committed:
Bump binary version to 1.33

Jefftree committed:
expand etcd storage tests to test for 1.31..1.33

Jefftree committed:
Expand emulated version range to n-3 with 1.31 floor

Jefftree commented on test/integration/etcd/data.go:
```diff
@@ -238,19 +238,19 @@ func GetEtcdStorageDataForNamespace(namespace string) map[schema.GroupVersionRes
-			ExpectedEtcdPath: "/registry/flowschemas/fs-2",
-			ExpectedGVK:      gvkP("flowcontrol.apiserver.k8s.io", "v1", "FlowSchema"),
-		},
+		// gvr("flowcontrol.apiserver.k8s.io", "v1beta3", "flowschemas"): {
```
Normally we can just remove these, but with emulated version we need to test the existence of these gvs on previous emulated versions. I don't have a good solution at the moment, but I think we need to version these fixtures to indicate the release they were added/removed similar to feature gates.

liggitt replied:
Don't comment this out, we need to find a way to fix the tests the 1.33 bump fails.

We already did this once in https://github.com/kubernetes/kubernetes/pull/126977/files (see the KUBE_APISERVER_SERVE_REMOVED_APIS_FOR_ONE_RELEASE=true additions)

liggitt replied:
These shouldn't be getting dropped in an uncontrolled way as part of this version bump, they should be getting dropped in a controlled way in a dedicated PR as we remove the serving code for each resource.

(I am in favor of dropping the serving code and cleaning these up, but not in this PR)

Jefftree replied:
Right, so we can't exactly use that env var again since it's only good for one release. I agree that commenting out isn't the correct solution and this is just the minimal set of changes to pass tests. 

`KUBE_APISERVER_SERVE_REMOVED_APIS_FOR_ONE_RELEASE=true` also only fixes the tests and doesn't delay any of the API removals scheduled to be removed at a particular release. All APIs designed to be removed at 1.32 were still removed at 1.32. If the types were removed in 1.32, we shouldn't exactly be testing them in 1.33?

liggitt replied:
> Right, so we can't exactly use that env var again since it's only good for one release.

It's not bounded to one release :)

Jefftree replied:
> It's not bounded to one release :)

The code looks to indicate it is only bound to one release, [link](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/server/deleted_kinds.go#L136), only reaching the serve for one more release condition if `removedVersion == currentVersion`.

> The types weren't removed, the serving code is still perfectly intact and can be activated with KUBE_APISERVER_SERVE_REMOVED_APIS_FOR_ONE_RELEASE=true. We can drop the tests as we drop the serving code, as in https://github.com/kubernetes/kubernetes/pull/129186

They can only be activated with `KUBE_APISERVER_SERVE_REMOVED_APIS_FOR_ONE_RELEASE=true` on 1.32 since the type was effectively removed in 1.32 (1.31 being the last release that they're served by default). I'm not aware of any way to activate it for 1.33 though I could be wrong.

There are currently two failing tests if we keep the fixtures: https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/128279/pull-kubernetes-integration/1867245952255922176. They both have `KUBE_APISERVER_SERVE_REMOVED_APIS_FOR_ONE_RELEASE=true` enabled and compare with discovery. We get an error: 

```
    /usr/local/google/home/jying/gospace/src/k8s.io/kubernetes/test/integration/etcd/etcd_storage_path_test.go:244: etcd data does not match the types we saw:
        in etcd data but not seen:
        [
        	admissionregistration.k8s.io/v1alpha1, Resource=validatingadmissionpolicies 
        	admissionregistration.k8s.io/v1alpha1, Resource=validatingadmissionpolicybindings 
        	flowcontrol.apiserver.k8s.io/v1beta3, Resource=flowschemas 
        	flowcontrol.apiserver.k8s.io/v1beta3, Resource=prioritylevelconfigurations]
        seen but not in etcd data:
        []
```

indicating that the resources are not in discovery.

I don't think it makes sense to test the supposedly removed resources on HEAD given they're removed in 1.32 and no cli option should allow them to be served on 1.33. Discussed with @jpbetz and we were thinking of creating a separate variable to record to set of resources removed in this release so we can fix the discovery and etcd fixture data diff error.

Jefftree commented on test/integration/apiserver/cel/admission_test_util.go:
```diff
@@ -137,17 +137,17 @@ var (
-		gvr("admissionregistration.k8s.io", "v1", "validatingadmissionpolicybindings"):        true,
-		gvr("admissionregistration.k8s.io", "v1alpha1", "mutatingadmissionpolicies"):          true,
-		gvr("admissionregistration.k8s.io", "v1alpha1", "mutatingadmissionpolicybindings"):    true,
+		// gvr("admissionregistration.k8s.io", "v1alpha1", "validatingadmissionpolicies"):        true,
```
Does CEL need to test on previous emulated versions? Can we remove these or do we need to keep them to test 1.32/1.31?

aojea commented:
/kind cleanuo
/priority important-soon

aojea commented:
/kind cleanup

aojea commented on test/integration/apiserver/cel/admission_policy_test.go:
```diff
@@ -535,13 +530,15 @@ func TestPolicyAdmission(t *testing.T) {
-		t.Fatal(err)
-	}
+	// 	t.Fatal(err)
+	// }
```
??

Jefftree replied:
oops, will clean this up

aojea commented:
/release-note none

BenTheElder commented on staging/src/k8s.io/component-base/version/version.go:
```diff
@@ -134,15 +134,15 @@ func (m *effectiveVersion) Validate() []error {
+	// emulationVersion can only be 1.{binaryMinor-2}...1.{binaryMinor}.
 	maxEmuVer := binaryVersion
-	minEmuVer := binaryVersion.SubtractMinor(1)
+	minEmuVer := binaryVersion.SubtractMinor(2)
```
(and in other places?)

Jefftree replied:
Opened https://github.com/kubernetes/kubernetes/pull/129208 for this

BenTheElder commented on test/integration/metrics/metrics_test.go:
```diff
@@ -119,7 +116,7 @@ func TestAPIServerMetrics(t *testing.T) {
 	}
 
 	// Make a request to a deprecated API to ensure there's at least one data point
```
... this seems rather fragile, but I guess we can't easily inject a fake deprecated API ...

Jefftree replied:
yeah I'll create an issue for this. Don't want to address it in this PR but always being forced to update it every couple of releases is not ideal.

BenTheElder replied:
Actually discussed briefly with @jpbetz that since we run api-server in-process we could probably ... actually register an integration-test-only deprecated API that will be perpetually deprecated?

BenTheElder commented:
> [do-not-merge] Test Binary Version Bump #128279

If you put [WIP] in the title then the CI will enforce "don't merge yet", though contributors can do this:
/retitle [WIP] [do-not-merge] Test Binary Version Bump

siyuanfoundation commented on staging/src/k8s.io/apiserver/pkg/server/options/server_run_options_test.go:
```diff
@@ -34,8 +34,8 @@ import (
 	featureGate := utilfeature.DefaultFeatureGate.DeepCopy()
-	effectiveVersion := utilversion.NewEffectiveVersion("1.30")
-	effectiveVersion.SetEmulationVersion(version.MajorMinor(1, 32))
+	effectiveVersion := utilversion.NewEffectiveVersion("1.32")
```
Like Ben suggested, we can subtract the real maximum and clamp it based on the lowest version to every support this (1.31), so max(binaryVersion.SubtractMinor(MAX_SKEW), v1.31.0), and set the test case to test for binary version 1.34, so that we don't need to change it going forward.

siyuanfoundation commented:
I think going forward, we can add an optional plow integration test to test for the next binary version, and this would send the signal to the developer that their test might break in the next version bump, and give hint if they want to fix the version in `StartTestServer` config.

liggitt commented on test/integration/apiserver/cel/admission_policy_test.go:
```diff
@@ -282,22 +281,22 @@ func createV1beta1ValidatingPolicyAndBinding(client clientset.Interface, convert
-	exact := admissionregistrationv1alpha1.Exact
-	equivalent := admissionregistrationv1alpha1.Equivalent
-	denyAction := admissionregistrationv1alpha1.DenyAction
+func createV1ValidatingPolicyAndBinding(client clientset.Interface, convertedRules []admissionregistrationv1.NamedRuleWithOperations) error {
```
dropping the alpha we will no longer serve is a reasonable thing to do, but doing that and making corresponding test changes should probably be done in a dedicated PR

Jefftree replied:
> dropping the alpha we will no longer serve is a reasonable thing to do, but doing that and making corresponding test changes should probably be done in a dedicated PR

ack

Jefftree replied:
Opened https://github.com/kubernetes/kubernetes/pull/129207

Jefftree replied:
https://github.com/kubernetes/kubernetes/pull/129207 has merged and rebased on this PR.

Jefftree commented:
I've isolated most of the commits into separate PRs per @liggitt 's suggestion. The main commit relevant for this PR is the first bump: https://github.com/kubernetes/kubernetes/pull/128279/commits/7c35c15b2e9a6663228a6e2d8007188f56f1a200
and etcd test change: https://github.com/kubernetes/kubernetes/pull/128279/commits/d5b0dcc05730272e60cd09736ac350e351eb7c07

Other PRs are covered by: https://github.com/kubernetes/kubernetes/pull/129186, https://github.com/kubernetes/kubernetes/pull/129207, https://github.com/kubernetes/kubernetes/pull/129208, https://github.com/kubernetes/kubernetes/pull/129207, https://github.com/kubernetes/kubernetes/pull/129211

siyuanfoundation commented on test/integration/metrics/metrics_test.go:
```diff
@@ -105,9 +105,6 @@ func TestAPIServerStorageMetrics(t *testing.T) {
-	// KUBE_APISERVER_SERVE_REMOVED_APIS_FOR_ONE_RELEASE allows for APIs pending removal to not block tests
-	t.Setenv("KUBE_APISERVER_SERVE_REMOVED_APIS_FOR_ONE_RELEASE", "true")
-
 	s := kubeapiservertesting.StartTestServerOrDie(t, nil, framework.DefaultTestServerFlags(), framework.SharedEtcd())
```
I would lock the binary version to 1.33 to prevent future failures 
```
&kubeapiservertesting.TestServerInstanceOptions{BinaryVersion: "1.32"}
```

It would work until AdmissionregistrationV1beta1 file is removed from the code base. Meanwhile we would have more time to work on a long term solution with fake api.

Jefftree replied:
Fixed

Jefftree committed:
remove emulated version test for un-emulatable 1.30

Jefftree commented:
API removal PRs have merged and I've updated this PR with the remaining fixes. PTAL Thanks!

A couple notes:
- Added an ignore list for etcd data tests that compare with discovery data. For APIs that are "removed" (unservable), the etcd data will be hidden. The APIs must remain because of API emulation, but on the latest release they are unservable.
- Removed a emulated version test that tests for 1.30. Unfortunately we don't have a GA API in the 1.31-1.32 range to test the storage version, it might be worthwhile to (similar to the metrics test) create a fake API that we inject into tests for this purpose.
- Changed the version validation to support n-3 with clamp at 1.31 min for emulatedVersion and 1.30 min for minCompatibilityVersion

liggitt commented on test/integration/metrics/metrics_test.go:
```diff
@@ -105,10 +105,7 @@ func TestAPIServerStorageMetrics(t *testing.T) {
-	t.Setenv("KUBE_APISERVER_SERVE_REMOVED_APIS_FOR_ONE_RELEASE", "true")
-
-	s := kubeapiservertesting.StartTestServerOrDie(t, nil, framework.DefaultTestServerFlags(), framework.SharedEtcd())
+	s := kubeapiservertesting.StartTestServerOrDie(t, &kubeapiservertesting.TestServerInstanceOptions{BinaryVersion: "1.33"}, framework.DefaultTestServerFlags(), framework.SharedEtcd())
```
What does pinning BinaryVersion to an old minor do? This seems pretty weird and likely to confuse / break various layers. Are you expecting to leave this pinned to 1.33, or is this another thing that would need bumping at start of 1.34 dev?

siyuanfoundation replied:
1.33 is pinned to the version `AdmissionregistrationV1beta1().ValidatingAdmissionPolicies` is deprecated. We have to pin the binary or emulation version so that we don't break the test once the version has moved across the border. Pinning emulation version would require we update this 3 versions later. Pinning the binary version would ensure this test works until the api is removed from the code base. 
We will work on a mock API to eventually replace this.

siyuanfoundation replied:
We can use the `--emulated-version` flag like in https://github.com/kubernetes/kubernetes/pull/129165. I think that would be better

Jefftree commented on test/integration/etcd/etcd_storage_path_test.go:
```diff
@@ -91,7 +91,7 @@ func TestEtcdStoragePath(t *testing.T) {
 	}
 
-	etcdStorageData := GetEtcdStorageData()
+	etcdStorageData := GetServedEtcdStorageDataForNamespace("etcdstoragepathtestnamespace")
```
We would need to version the etcd data fixtures for both when a type was added and removed to properly support emulation version. Is this something that should apply only to the etcd storage tests or all tests that use this function? Things like encryption, field validation, etc all use this function. Should all callers run the suite of tests on supported emulation versions?

jpbetz commented on staging/src/k8s.io/component-base/version/version.go:
```diff
@@ -134,15 +136,15 @@ func (m *effectiveVersion) Validate() []error {
+	// emulationVersion can only be 1.{binaryMinor-3}...1.{binaryMinor}
 	maxEmuVer := binaryVersion
-	minEmuVer := binaryVersion.SubtractMinor(1)
+	minEmuVer := binaryVersion.SubtractMinor(3)
```
Can we open an issue to move this logic out of component-base as well?  I would expect k8s components (kube-apiserver, ...) to set the limit according to the rules of that component (N-3 for k8s components), but leave it up to 3rd party component owners to decide what the limit should be for their components.

siyuanfoundation replied:
That's a good idea. Would `staging/src/k8s.io/apimachinery/pkg/version` be a reasonable place to put it?

jpbetz replied:
Yes, and this PR doesn't need to address that problem now since it's preexisting, it just highlights that we should track and fix in a follow up

Jefftree replied:
Opened https://github.com/kubernetes/kubernetes/issues/129291 as a follow up

Jefftree commented on staging/src/k8s.io/component-base/version/version_test.go:
```diff
@@ -64,19 +77,27 @@ func TestValidate(t *testing.T) {
-			emulationVersion:        "v1.32.0",
-			minCompatibilityVersion: "v1.30.0",
-			expectErrors:            true,
+			name:                    "compatibility version two minor lower than binary ok",
```
Fixed

Jefftree commented on staging/src/k8s.io/component-base/version/version.go:
```diff
@@ -25,6 +25,8 @@ import (
 	apimachineryversion "k8s.io/apimachinery/pkg/version"
 )
 
+var floorVersion *version.Version = version.MajorMinor(1, 31)
```
cc @jpbetz @siyuanfoundation Similar comment: https://github.com/kubernetes/kubernetes/pull/128279/files#r1890662857

I think it makes sense to move the validation out of component base since this is shared.

Opened https://github.com/kubernetes/kubernetes/issues/129291

liggitt replied:
at least rename this to minimumKubeEmulationVersion to make it clear this is kube-specific and bounds the emulation version

Jefftree replied:
Renamed.

siyuanfoundation commented on staging/src/k8s.io/component-base/version/version.go:
```diff
@@ -134,15 +136,15 @@ func (m *effectiveVersion) Validate() []error {
-	maxCompVer := binaryVersion.SubtractMinor(1)
-	minCompVer := binaryVersion.SubtractMinor(1)
+	// minCompatibilityVersion can only be 1.{binaryMinor-3} to 1.{binaryMinor}
+	maxCompVer := binaryVersion
+	minCompVer := binaryVersion.SubtractMinor(3)
```
```suggestion
	maxCompVer := emulationVersion
	minCompVer := binaryVersion.SubtractMinor(4)
```

siyuanfoundation replied:
When `emulationVersion  = binaryVersion - 3`, we still need to support `minCompatibilityVersion =emulationVersion-1`

siyuanfoundation replied:
Can you also add 
```diff

func (m *effectiveVersion) SetEmulationVersion(emulationVersion *version.Version) {
	m.emulationVersion.Store(majorMinor(emulationVersion))
+.     // set the default minCompatibilityVersion to be emulationVersion - 1
+      m.minCompatibilityVersion.Store(majorMinor(minCompatibilityVersion.SubtractMinor(1)))
}

+ // SetMinCompatibilityVersion should be called after SetEmulationVersion
func (m *effectiveVersion) SetMinCompatibilityVersion(minCompatibilityVersion *version.Version) {
	m.minCompatibilityVersion.Store(majorMinor(minCompatibilityVersion))
}
?
```

Jefftree replied:
Fixed

Jefftree commented on test/integration/controlplane/transformation/kms_transformation_test.go:
```diff
@@ -644,7 +644,7 @@ resources:
 		client := dynamic.NewForConfigOrDie(test.kubeAPIServer.ClientConfig)
 
-		etcdStorageData := etcd.GetEtcdStorageDataForNamespace(testNamespace)
+		etcdStorageData := etcd.GetServedEtcdStorageDataForNamespace(testNamespace)
```
Added tests to exercise different emulated versions, changed quite a few files though.

Jefftree commented on test/integration/apiserver/apiserver_test.go:
```diff
@@ -2975,20 +2975,6 @@ func TestEmulatedStorageVersion(t *testing.T) {
 		expectedStorageVersion schema.GroupVersion
 	}
 	cases := []testCase{
-		{
```
cc @siyuanfoundation: We're losing coverage of the GA API storage version transition and we do not have a GA API that could be used for 1.31 and 1.32. We can test another beta API or maybe figure out a way to inject fake one?

jpbetz replied:
+1 to keeping this test.

Since we don't have a existing build in that meets the criteria to be used in the test case.  This might be a good time to figure out a more sustainable solution (xref https://github.com/kubernetes/kubernetes/pull/128279/files?diff=unified&w=1#r1882520074 which as the same need).

@liggitt unless there is an easy way for us to create a fake API type that we can use here.  How about we open an issue to sort this out as a follow up?  That way we can get this bump merged ASAP, and then fix this test coverage the "right way".  We can promise to have something in place before 1.33 enhancement freeze.

jpbetz replied:
I've dropped some notes on how we might follow up here: https://github.com/kubernetes/kubernetes/issues/129311

siyuanfoundation commented on test/integration/etcd/etcd_storage_path_test.go:
```diff
@@ -75,197 +77,205 @@ var allowMissingTestdataFixtures = map[schema.GroupVersionKind]bool{
+						parts := strings.Split(filepath.Base(previousReleaseGroupKindFile), ".")
+						version := parts[len(parts)-3]
+						if !strings.Contains(version, "alpha") {
+							if serverVersion := filepath.Base(filepath.Dir(previousReleaseGroupKindFile)); serverVersion == "HEAD" {
```
the current and previous version should be based on emulation version.

aojea commented:
> Kubernetes e2e suite: [It] [sig-cli] Kubectl Port forwarding Shutdown client connection while the remote stream is writing data to the port-forward connection port-forward should keep working after detect broken connection expand_more	

seems unrelated to this PR but not a flake we like to see cc: @soltysh

soltysh commented:
> https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/128279/pull-kubernetes-e2e-gce/1869835464060440576
> seems unrelated to this PR but not a flake we like to see cc: @soltysh

@aojea we tracked this particular one already back [here](https://github.com/kubernetes/kubernetes/pull/128319#issuecomment-2444602885) and we agreed that bumping this particular timeout to 10s should do the trick. If the problem re-appears, and given it only applies to gce we can bump this a bit more.

liggitt commented on test/integration/metrics/metrics_test.go:
```diff
@@ -105,10 +105,7 @@ func TestAPIServerStorageMetrics(t *testing.T) {
 
 func TestAPIServerMetrics(t *testing.T) {
-	// KUBE_APISERVER_SERVE_REMOVED_APIS_FOR_ONE_RELEASE allows for APIs pending removal to not block tests
-	t.Setenv("KUBE_APISERVER_SERVE_REMOVED_APIS_FOR_ONE_RELEASE", "true")
```
leave this bit as-is, it doesn't hurt anything and we can clean it up when we stop using a real beta type for this test

liggitt commented on staging/src/k8s.io/component-base/featuregate/registry_test.go:
```diff
@@ -125,7 +125,7 @@ func TestVersionFlagOptionsWithMapping(t *testing.T) {
 	}
 	minCompVers := strings.Join(r.unsafeVersionFlagOptions(false), "\n")
-	expectedMinCompVers := "kube=1.30..1.31 (default=1.30)\ntest=2.7..2.8 (default=2.7)"
+	expectedMinCompVers := "test=2.7..2.8 (default=2.7)"
```
it's surprising the kube portion dropped from the message... is that intentional?

Jefftree replied:
It was not intentional. This has been updated: https://github.com/kubernetes/kubernetes/pull/128279#discussion_r1894241284

liggitt commented on test/integration/etcd/data.go:
any API objects which *change* their storage version depending on emulation version will need to calculate their expected GVK from the incoming emulationVersion as well, for example:

```
			ExpectedGVK:      (func() *schema.GroupVersionKind {
				switch emulationVersion {
				case "1.33":
					return gvkP("...."),
				case "1.32":
					return gvkP("...."),
				default:
					return gvkP("...."),
				}
			})(),
```

Jefftree replied:
> any API objects which _change_ their storage version depending on emulation version will need to calculate their expected GVK from the incoming emulationVersion as well, for example:
> 
> ```
> 			ExpectedGVK:      (func() *schema.GroupVersionKind {
> 				switch emulationVersion {
> 				case "1.31":
> 					return gvkP("...."),
> 				case "1.32":
> 					return gvkP("...."),
> 				default:
> 					return gvkP("...."),
> 				}
> 			})(),
> ```

ack, we do not have any instances of this at the moment and it'll probably occur on our next GA API promotion. 

I've addressed the other diffs.

Jefftree committed:
fix metrics test

Jefftree commented on staging/src/k8s.io/component-base/featuregate/registry_test.go:
```diff
@@ -118,7 +118,7 @@ func TestVersionFlagOptions(t *testing.T) {
 	r := testRegistry(t)
 	utilruntime.Must(r.SetEmulationVersionMapping(testComponent, DefaultKubeComponent,
-		func(from *version.Version) *version.Version { return from.OffsetMinor(3) }))
+		func(from *version.Version) *version.Version { return version.MajorMinor(1, from.Minor()+23) }))
```
This was updated because the original test was not written to check the actual values of the mapping, just a placeholder for a mapping exist. So it is actually mapping test 2.8 to kube 2.11, which is not aligned with the kube binary at all.

This change fixes the test to the correct version mapping. 

cc @siyuanfoundation @liggitt

liggitt replied:
is adding 23 to the minor correct?

Jefftree replied:
Mapping is from test component 2.7 -> kube 1.30 and 2.8 -> kube 1.31 which is where the +23 was derived from specifically for this test.

Jefftree reviewed (commented):
Updated PTAL

Jefftree commented on test/integration/etcd/server.go:
```diff
@@ -62,6 +67,17 @@ AwEHoUQDQgAEH6cuzP8XuD5wal6wf9M6xDljTOPLX2i8uIp/C/ASqiIGUeeKQtX0
 /IR3qCXyThP/dbCiHrF3v1cuhBOHY8CLVg==
 -----END EC PRIVATE KEY-----`
 
+func registerEffectiveEmulationVersion(t *testing.T) {
```
@liggitt: Unfortunately I couldn't do this as part of `configFuncs` as this must be called before we create ` options.NewServerRunOptions()`. It locks the registered version https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/server/options/server_run_options.go#L104-L108. cc @siyuanfoundation

liggitt replied:
ok, if it becomes conditional on an emulatedVersion passed into the server start method, which only the one test we care about sets, that's ok as well

Jefftree replied:
Yes, this is only called on the one test we care about: https://github.com/kubernetes/kubernetes/blob/c62628e520f576162e6836debd81f5c3ee767450/test/integration/etcd/etcd_storage_path_test.go#L95

Jefftree commented on test/integration/etcd/data.go:
```diff
@@ -32,6 +32,44 @@ func GetEtcdStorageData() map[schema.GroupVersionResource]StorageData {
+func GetRemovedEtcdStorageDataForNamespace(namespace string) map[schema.GroupVersionResource]StorageData {
+	return map[schema.GroupVersionResource]StorageData{
+		// k8s.io/kubernetes/pkg/apis/flowcontrol/v1beta3
+		gvr("flowcontrol.apiserver.k8s.io", "v1beta3", "flowschemas"): {
```
This has been addressed.

liggitt commented on staging/src/k8s.io/component-base/version/version.go:
```diff
@@ -134,15 +139,15 @@ func (m *effectiveVersion) Validate() []error {
-	maxCompVer := binaryVersion.SubtractMinor(1)
-	minCompVer := binaryVersion.SubtractMinor(1)
+	// minCompatibilityVersion can only be 1.{binaryMinor-3} to 1.{binaryMinor}
+	maxCompVer := emulationVersion
```
this is letting the compatibility version be == binary version? that would mean we'd allow a cluster to start on 1.33 and not be compatible with a 1.32 apiserver... are we really intending to relax that way in this PR?

siyuanfoundation replied:
minCompVer is not exposed with any flag yet, and it stays at the default emu-1. Eventually we want to support compatibility version be == binary version. This is here so that we do not need to change it later.

Jefftree replied:
This was a bug. Per the [KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-architecture/4330-compatibility-versions/README.md#motivation)

> Lastly, a --min-compatibility-version can be set to the binary version for new clusters being used for green field projects, providing immediate access to the latest Kubernetes features without the need to wait an additional release for features to settle in as is typically needed for rollback support.

Default is still at emulationVersion - 1.

liggitt replied:
I know this was the aspirational end state for the KEP, but I didn't know if we were ready to relax to that yet. If it's not wired to a flag yet so there's no way to change it from emulationVersion-1 at this point, that's fine.

liggitt commented on test/integration/etcd/etcd_storage_path_test.go:
```diff
@@ -75,13 +78,30 @@ var allowMissingTestdataFixtures = map[schema.GroupVersionKind]bool{
-	t.Setenv("KUBE_APISERVER_SERVE_REMOVED_APIS_FOR_ONE_RELEASE", "true")
+	supportedVersions := GetSupportedEmulatedVersions()
+	for _, v := range supportedVersions {
+		testEtcdStoragePathWithVersion(t, v)
```
nit: make a subtest to separate results
```suggestion
		t.Run(v, func(t *testing.T) {
			testEtcdStoragePathWithVersion(t, v)
		})
```

Jefftree replied:
Updated

liggitt commented on test/integration/etcd/etcd_storage_path_test.go:
```diff
@@ -75,13 +78,32 @@ var allowMissingTestdataFixtures = map[schema.GroupVersionKind]bool{
 // It will also fail when a type gets moved to a different location. Be very careful in this situation because
 // it essentially means that you will be break old clusters unless you create some migration path for the old data.
 func TestEtcdStoragePath(t *testing.T) {
```
Functionally, this test change lgtm. Check how long this package and this test takes to run in CI now. I want to make sure we're not risking getting close to package timeout limits.

Jefftree replied:
From the [latest run on master](https://storage.googleapis.com/kubernetes-ci-logs/logs/ci-kubernetes-integration-master/1870168235089858560/build-log.txt): 

> ok  	k8s.io/kubernetes/test/integration/etcd	103.141s

From the integration run on this PR:
https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/128279/pull-kubernetes-integration/1870184749037785088

> ok  	k8s.io/kubernetes/test/integration/etcd	119.389s

Jefftree commented:
```
k8s.io/client-go/tools: cache expand_less | 44s
{Failed  === RUN   TestAddWhileActive
polling
==================
WARNING: DATA RACE
```

This seems a bit worrisome but I was also able to reproduce this on master. Will file an issue

/test pull-kubernetes-unit

liggitt commented:
/lgtm
/approve

aojea commented on test/integration/etcd/etcd_storage_path_test.go:
```diff
@@ -75,13 +78,32 @@ var allowMissingTestdataFixtures = map[schema.GroupVersionKind]bool{
+		// Only test for beta and GA APIs with emulated version.
+		featuregatetesting.SetFeatureGateEmulationVersionDuringTest(t, feature.DefaultFeatureGate, version.MustParse(v))
+		featuregatetesting.SetFeatureGateDuringTest(t, feature.DefaultFeatureGate, "AllBeta", true)
+		registerEffectiveEmulationVersion(t)
```
This presents a problem , and it took me several days to find it but I still can't figure out how to solve it.

If you checkout my PR https://github.com/kubernetes/kubernetes/pull/128971 and apply the following patch

```diff
diff --git a/pkg/controlplane/apiserver/apis.go b/pkg/controlplane/apiserver/apis.go
index 7ed3a58489b..47dc2cf2467 100644
--- a/pkg/controlplane/apiserver/apis.go
+++ b/pkg/controlplane/apiserver/apis.go
@@ -89,6 +89,7 @@ func (s *Server) InstallAPIs(restStorageProviders ...RESTStorageProvider) error
        nonLegacy := []*genericapiserver.APIGroupInfo{}
 
        // used later in the loop to filter the served resource by those that have expired.
+       klog.Infof("DEBUG ---------- effective emulation version %s", s.GenericAPIServer.EffectiveVersion.EmulationVersion().String())
        resourceExpirationEvaluator, err := genericapiserver.NewResourceExpirationEvaluator(s.GenericAPIServer.EffectiveVersion.EmulationVersion())
        if err != nil {
                return err
```
and runs the test
```
go test -timeout 600s -run ^TestEtcdStoragePath$ k8s.io/kubernetes/test/integration/etcd -v -count 1 | grep -E 'DEBUG|TestEtcdStoragePath'
```

you can see that the 1.33 fails because it it keeps using the emulated version 1.32 with 1.33
```
=== RUN   TestEtcdStoragePath/1.33
I0109 00:21:01.588321 2218629 serving.go:380] Generated self-signed cert (/tmp/TestEtcdStoragePath1.3382862738/apiserver.crt, /tmp/TestEtcdStoragePath1.3382862738/apiserver.key)
I0109 00:21:01.590392 2218629 dynamic_serving_content.go:116] "Loaded a new cert/key pair" name="serving-cert::/tmp/TestEtcdStoragePath1.3382862738/apiserver.crt::/tmp/TestEtcdStoragePath1.3382862738/apiserver.key"
I0109 00:21:01.989808 2218629 apf_controller.go:292] NewTestableController "Controller" with serverConcurrencyLimit=600, name=Controller, asFieldManager="api-priority-and-fairness-config-consumer-v1"
I0109 00:21:02.018277 2218629 apis.go:92] DEBUG ---------- effective emulation version 1.32
```

however, if you run the test individually it works fine

```
PASS
{"level":"warn","ts":"2025-01-09T00:24:46.595770Z","caller":"embed/serve.go:160","msg":"stopping insecure grpc server due to error","error":"accept tcp 127.0.0.1:42889: use of closed network connection"}
{"level":"warn","ts":"2025-01-09T00:24:46.595848Z","caller":"embed/serve.go:162","msg":"stopped insecure grpc server due to error","error":"accept tcp 127.0.0.1:42889: use of closed network connection"}
I0109 00:24:51.596565 2223139 etcd.go:158] etcd didn't exit in 5 seconds, killing it
I0109 00:24:51.596601 2223139 etcd.go:164] etcd exit status: signal: terminated
ok      k8s.io/kubernetes/test/integration/etcd 19.938s
```

cc: @liggitt @Jefftree @siyuanfoundation

Jefftree replied:
Confirming this is an issue. The variable is actually reset properly but we need to call https://github.com/kubernetes/kubernetes/blob/master/cmd/kube-apiserver/app/testing/testserver.go#L208-L209 to register the proper effective version with the component. Will send a fix

siyuanfoundation replied:
Sorry about the frustration. https://github.com/kubernetes/kubernetes/pull/129416 should be able to fix the problem.

aojea replied:
> Sorry about the frustration. #129416 should be able to fix the problem.

not sorry about nothing, these things happen, we need to break some eggs to do the omelette :)
